app/gui2/src/util/ffi.ts

```diff
@@ -0,0 +1,406 @@
+    export type Wildcard = TokenBase<{
+      lift_level: number
+    }>
+  }
+  export interface Span {
+    code_length: Length
+    left_offset: Offset
+  }
+
+  export interface Code {
+    repr: CowStrPtr
+    utf16: number
+  }
+
+  export interface Length {
+    utf8: number
+    utf16: number
+  }
+
+  export interface Offset {
+    visible: VisibleOffset
+    code: Code
+  }
+
+  export interface CowStrPtr {
+    begin: number
+    len: number
+  }
+
+  export interface VisibleOffset {
+    width_in_spaces: number
+  }
+
+  export type Variant =
+    | { Invalid: Variant.Invalid }
+    | { BodyBlock: Variant.BodyBlock }
+    | { ArgumentBlockApplication: Variant.ArgumentBlockApplication }
+    | { OperatorBlockApplication: Variant.OperatorBlockApplication }
+    | { Ident: Variant.Ident }
+    | { Number: Variant.Number }
+    | { Wildcard: Variant.Wildcard }
+    | { AutoScope: Variant.AutoScope }
+    | { TextLiteral: Variant.TextLiteral }
+    | { App: Variant.App }
+    | { NamedApp: Variant.NamedApp }
+    | { DefaultApp: Variant.DefaultApp }
+    | { OprApp: Variant.OprApp }
+    | { UnaryOprApp: Variant.UnaryOprApp }
+    | { OprSectionBoundary: Variant.OprSectionBoundary }
+    | { TemplateFunction: Variant.TemplateFunction }
+    | { MultiSegmentApp: Variant.MultiSegmentApp }
+    | { TypeDef: Variant.TypeDef }
+    | { Assignment: Variant.Assignment }
+    | { Function: Variant.Function }
+    | { ForeignFunction: Variant.ForeignFunction }
+    | { Import: Variant.Import }
+    | { Export: Variant.Export }
+    | { Group: Variant.Group }
+    | { TypeSignature: Variant.TypeSignature }
+    | { TypeAnnotated: Variant.TypeAnnotated }
+    | { CaseOf: Variant.CaseOf }
+    | { Lambda: Variant.Lambda }
+    | { Array: Variant.Array }
+    | { Tuple: Variant.Tuple }
+    | { Annotated: Variant.Annotated }
+    | { AnnotatedBuiltin: Variant.AnnotatedBuiltin }
+    | { Documented: Variant.Documented }
+    | { ConstructorDefinition: Variant.ConstructorDefinition }
+
+  module Variant {
+    export interface Invalid {
+      error: Error
+      ast: Tree
+    }
+
+    export interface BlockLine {
+      newline: Token.Newline
+      expression: Opt<Tree>
+    }
+
+    export interface BlockOperatorLine {}
+
+    type OperatorOrError = { Ok: Token } | { Err: Error }
+    export interface BodyBlock {
+      statements: BlockLine[]
+    }
+
+    export interface ArgumentBlockApplication {
+      lhs: Opt<Tree>
+      arguments: BlockLine[]
+    }
+    export interface OperatorBlockApplication {
+      lhs: Opt<Tree>
+      expressions: BlockOperatorLine[]
+      excess: BlockLine[]
+    }
+    export interface Ident {
+      token: Token.Ident
+    }
+
+    export interface FractionalDigits {
+      dot: Token.Operator
+      digits: Token.Digits
+    }
+
+    export interface Number {
+      base: Opt<Token.NumberBase>
+      integer: Opt<Token.Digits>
+      fractional_digits: Opt<FractionalDigits>
```

Comment:
a bit surprised rust doesnt `#[serde(rename_all = "camelCase")]` but i guess it's fine?

Reply:
Yeah, I don't want to mess with the parser codebase too much, as we also generate Java code from it for the engine. Therefore the existing struct field names are all effectively public API.